### PR TITLE
CoreData improvements

### PR DESCRIPTION
**Pull request by tjprescott**
Changes to improve the CoreData stability and performance, as well as some problems with the UI that were bogging down performance (and may have been responsible for some of the issues)

A transfer print-trace looks like:
```
RELOAD TABLE
1 didUpdate
RELOAD TABLE
Transfer 2279286792 didUpdateWithState: inProgress andProgress: 0.055555556
RELOAD TABLE
Transfer 2279286792 didUpdateWithState: inProgress andProgress: 0.11111111
RELOAD TABLE
Transfer 2279286792 didUpdateWithState: inProgress andProgress: 0.16666667
RELOAD TABLE
Transfer 2279286792 didUpdateWithState: inProgress andProgress: 0.22222222
RELOAD TABLE
Transfer 2279286792 didUpdateWithState: inProgress andProgress: 0.2777778
RELOAD TABLE
Transfer 2279286792 didUpdateWithState: inProgress andProgress: 0.33333334
RELOAD TABLE
Transfer 2279286792 didUpdateWithState: inProgress andProgress: 0.3888889
RELOAD TABLE
Transfer 2279286792 didUpdateWithState: inProgress andProgress: 0.44444445
RELOAD TABLE
Transfer 2279286792 didUpdateWithState: inProgress andProgress: 0.5
RELOAD TABLE
Transfer 2279286792 didUpdateWithState: inProgress andProgress: 0.5555556
RELOAD TABLE
Transfer 2279286792 didUpdateWithState: inProgress andProgress: 0.6111111
RELOAD TABLE
Transfer 2279286792 didUpdateWithState: inProgress andProgress: 0.6666667
RELOAD TABLE
Transfer 2279286792 didUpdateWithState: inProgress andProgress: 0.7222222
RELOAD TABLE
Transfer 2279286792 didUpdateWithState: inProgress andProgress: 0.7777778
RELOAD TABLE
Transfer 2279286792 didUpdateWithState: inProgress andProgress: 0.8333333
RELOAD TABLE
Transfer 2279286792 didUpdateWithState: inProgress andProgress: 0.8888889
RELOAD TABLE
Transfer 2279286792 didUpdateWithState: inProgress andProgress: 0.9444444
RELOAD TABLE
Transfer 2279286792 didUpdateWithState: inProgress andProgress: 1.0
RELOAD TABLE
Transfer 2279286792 didComplete
RELOAD TABLE
```

Prior to this, the table would _continuously_ reload, and delegates were being signaled for blocks (regression from a rebase, I think).

### CODE REVIEW

File: examples/AzureSDKDemoSwift/Source/BlobDownloadViewController.swift [hunk 1]
```diff
@@ -90,8 +90,12 @@ class BlobDownloadViewController: UIViewController, MSALInteractiveDelegate {
...
         dataSource?.nextPage { result in
             switch result {
-            case .success:
-                self.reloadTableView()
+            case let .success(data):
+                if data != nil {
+                    self.reloadTableView()
+                } else {
+                    self.noMoreData = true
+                }
```
Comment (tjprescott) [inline examples/AzureSDKDemoSwift/Source/BlobDownloadViewController.swift:98]: This was causing a continuous signal to reload the table view.

File: examples/AzureSDKDemoSwift/Source/BlobDownloadViewController.swift [hunk 2]
```diff
@@ -102,6 +106,7 @@ class BlobDownloadViewController: UIViewController, MSALInteractiveDelegate {
     /// Reload the table view on the UI thread.
     private func reloadTableView() {
         DispatchQueue.main.async { [weak self] in
+            print("RELOAD TABLE")
```
Comment (tjprescott) [inline examples/AzureSDKDemoSwift/Source/BlobDownloadViewController.swift:109]: I'll remove these print statements before merging.

File: examples/AzureSDKDemoSwift/Source/BlobDownloadViewController.swift [hunk 3]
```diff
@@ -228,20 +232,23 @@ extension BlobDownloadViewController: TransferDelegate {
 
     func transfer(
         _ transfer: Transfer,
-        didUpdateWithState _: TransferState,
+        didUpdateWithState state: TransferState,
         andProgress progress: Float?
     ) {
         if let blobTransfer = transfer as? BlobTransfer, blobTransfer.transferType == .download {
+            print("Transfer \(blobTransfer.hash) didUpdateWithState: \(state) andProgress: \(progress ?? -1)")
```
Comment (tjprescott) [inline examples/AzureSDKDemoSwift/Source/BlobDownloadViewController.swift:239]: Same here.

File: examples/AzureSDKDemoSwift/Source/BlobDownloadViewController.swift [hunk 4]
```diff
@@ -228,20 +232,23 @@ extension BlobDownloadViewController: TransferDelegate {
...
             reloadTableView()
         }
     }
 
-    func transfersDidUpdate(_: [Transfer]) {
+    func transfersDidUpdate(_ transfers: [Transfer]) {
+        print("\(transfers.count) didUpdate")
```
Comment (tjprescott) [inline examples/AzureSDKDemoSwift/Source/BlobDownloadViewController.swift:245]: Same here.

File: examples/AzureSDKDemoSwift/Source/BlobDownloadViewController.swift [hunk 5]
```diff
@@ -228,20 +232,23 @@ extension BlobDownloadViewController: TransferDelegate {
...
         reloadTableView()
     }
 
     func transferDidComplete(_ transfer: Transfer) {
         if let blobTransfer = transfer as? BlobTransfer, blobTransfer.transferType == .download {
+            print("Transfer \(blobTransfer.hash) didComplete")
```
Comment (tjprescott) [inline examples/AzureSDKDemoSwift/Source/BlobDownloadViewController.swift:251]: Same here.

File: sdk/storage/AzureStorageBlob/Source/Data/AzureStorage.xcdatamodeld/AzureStorage.xcdatamodel/contents [hunk 6]
```diff
@@ -23,7 +23,7 @@
         <attribute name="id" optional="YES" attributeType="UUID" usesScalarValueType="NO"/>
         <attribute name="rawState" optional="YES" attributeType="Integer 16" defaultValueString="0" usesScalarValueType="YES"/>
         <attribute name="startRange" attributeType="Integer 64" defaultValueString="0" usesScalarValueType="YES"/>
-        <relationship name="parent" optional="YES" maxCount="1" deletionRule="Nullify" destinationEntity="BlobTransfer" inverseName="blocks" inverseEntity="BlobTransfer"/>
+        <relationship name="parent" maxCount="1" deletionRule="Nullify" destinationEntity="BlobTransfer" inverseName="blocks" inverseEntity="BlobTransfer"/>
```
Comment (tjprescott) [inline sdk/storage/AzureStorageBlob/Source/Data/AzureStorage.xcdatamodeld/AzureStorage.xcdatamodel/contents:26]: `BlockTransfer.parent` is no longer optional.

File: sdk/storage/AzureStorageBlob/Source/Operation/TransferOperation.swift [hunk 7]
```diff
@@ -51,26 +51,15 @@ internal class TransferOperation: Operation {
     // MARK: Internal Methods
 
     internal func notifyDelegate(withTransfer transfer: Transfer) {
```
Comment (tjprescott) [inline sdk/storage/AzureStorageBlob/Source/Operation/TransferOperation.swift:53]: Ensure that the BlobTransfer is the only thing signaled by the delegate.

File: sdk/storage/AzureStorageBlob/Source/Transport/URLSessionTransferManager.swift [hunk 8]
```diff
@@ -75,16 +75,26 @@ internal final class URLSessionTransferManager: NSObject, TransferManager, URLSe
...
 
-    var persistentContainer: NSPersistentContainer? = {
-        guard let bundle = Bundle(identifier: "com.azure.storage.AzureStorageBlob") else { return nil }
-        guard let url = bundle.url(forResource: "AzureStorage", withExtension: "momd") else { return nil }
-        guard let model = NSManagedObjectModel(contentsOf: url) else { return nil }
+    var persistentContainer: NSPersistentContainer {
+        return URLSessionTransferManager.sharedPersistentContainer
+    }
+
+    static var sharedPersistentContainer: NSPersistentContainer = {
```
Comment (tjprescott) [inline sdk/storage/AzureStorageBlob/Source/Transport/URLSessionTransferManager.swift:82]: This would be groundwork to allow us to instantiate different TransferManagers.

File: examples/AzureSDKDemoSwift/Source/BlobDownloadViewController.swift (hunk 2 above)
Comment (tjprescott) [reply examples/AzureSDKDemoSwift/Source/BlobDownloadViewController.swift]: Removed.

File: sdk/storage/AzureStorageBlob/Source/Operation/TransferOperationQueue.swift [hunk 9]
```diff
@@ -34,27 +34,15 @@ internal class TransferOperationQueue: OperationQueue {
...
-                    let operationClosure = operation.completionBlock
-                    operation.completionBlock = {
-                        operationClosure?()
-                        if !operation.transfer.isActive {
-                            return
-                        }
-                    }
-                }
-                addOperations(filteredOps, waitUntilFinished: false)
+        let filteredOps = operations.filter { $0.transfer.state.active }
```
Comment (bsiegel) [inline sdk/storage/AzureStorageBlob/Source/Operation/TransferOperationQueue.swift:37]: Is this the correct filter? What states must a transfer be in, for them to be eligible to be added to the operation queue? My guess would be pending, failed, and paused.

File: sdk/storage/AzureStorageBlob/Source/StorageBlobClient.swift [hunk 10]
```diff
@@ -63,6 +64,12 @@ public final class StorageBlobClient: PipelineClient {
 
     private static let manager = URLSessionTransferManager.shared
 
+    internal static let viewContext: NSManagedObjectContext = manager.persistentContainer.viewContext
+
+    internal static func newBackgroundContext() -> NSManagedObjectContext {
```
Comment (bsiegel) [inline sdk/storage/AzureStorageBlob/Source/StorageBlobClient.swift:69]: Doesn't look like this is ever called?

File: sdk/storage/AzureStorageBlob/Source/Transport/URLSessionTransferManager+TransferDelegate.swift [hunk 11]
```diff
@@ -59,34 +62,42 @@ extension URLSessionTransferManager: TransferDelegate {
...
+        }
     }
 
     func transfersDidUpdate(_ transfers: [Transfer]) {
-        let restorationIds = transfers.compactMap { ($0 as? TransferImpl)?.clientRestorationId }
-        for restorationId in Set(restorationIds) {
-            guard let delegate = client(forRestorationId: restorationId) as? TransferDelegate else { continue }
-            let transfersForId = transfers.filter { ($0 as? TransferImpl)?.clientRestorationId == restorationId }
-            delegate.transfersDidUpdate(transfersForId)
+        guard let restorationId = (transfers.first as? TransferImpl)?.clientRestorationId else { return }
```
Comment (bsiegel) [inline sdk/storage/AzureStorageBlob/Source/Transport/URLSessionTransferManager+TransferDelegate.swift:74]: Is this a regression, or are we sure that all transfers will share the same restoration ID?

File: sdk/storage/AzureStorageBlob/Source/Transport/URLSessionTransferManager.swift [hunk 12]
```diff
@@ -387,7 +395,7 @@ internal final class URLSessionTransferManager: NSObject, TransferManager, URLSe
     }
 
     func pause(transfer: BlobTransfer) {
-        guard transfer.state.active else { return }
```
Comment (bsiegel) [inline sdk/storage/AzureStorageBlob/Source/Transport/URLSessionTransferManager.swift:390]: Why this change? It doesn't seem like calling `pause` on a paused or failed transfer makes sense.

File: sdk/storage/AzureStorageBlob/Source/Transport/URLSessionTransferManager+TransferDelegate.swift (hunk 11 above)
Comment (tjprescott) [reply sdk/storage/AzureStorageBlob/Source/Transport/URLSessionTransferManager+TransferDelegate.swift:74]: I think probably a bad rebase. We don't want to assume all restoration IDs are the same.

File: sdk/storage/AzureStorageBlob/Source/StorageBlobClient.swift (hunk 10 above)
Comment (tjprescott) [reply sdk/storage/AzureStorageBlob/Source/StorageBlobClient.swift:69]: It isn't currently called because once I enabled concurrency debugging to begin using this it was clear that should be a separate PR.

File: sdk/storage/AzureStorageBlob/Source/Operation/TransferOperationQueue.swift (hunk 9 above)
Comment (tjprescott) [reply sdk/storage/AzureStorageBlob/Source/Operation/TransferOperationQueue.swift:37]: According to the prior logic, only `.pending` and `.inProgress` transfers would be permitted. This simplifies that to use `state.active`. For failed and paused operations, they would be put into pending state prior to this method being called.

File: sdk/storage/AzureStorageBlob/Source/StorageBlobClient.swift (hunk 10 above)
Comment (tjprescott) [reply sdk/storage/AzureStorageBlob/Source/StorageBlobClient.swift:69]: I've removed it for now.

File: sdk/storage/AzureStorageBlob/Source/Transport/URLSessionTransferManager.swift (hunk 12 above)
Comment (tjprescott) [reply sdk/storage/AzureStorageBlob/Source/Transport/URLSessionTransferManager.swift:390]: Can't recall offhand. Switched it back to `.active`.

File: sdk/storage/AzureStorageBlob/Source/Transport/URLSessionTransferManager+TransferDelegate.swift (hunk 11 above)
Comment (tjprescott) [reply sdk/storage/AzureStorageBlob/Source/Transport/URLSessionTransferManager+TransferDelegate.swift]: Fixed
